### PR TITLE
feat: support `equal` for length and range rules

### LINKED ISSUE #41: equal
equal for length like in the `validator` crate

### PR DESCRIPTION
**Pull request by Rolv-Apneseth**
- Closes #41

Hey, just trying to tackle #41 

I'm a bit out of my depth here so please let me know if I missed something / did something wrong.

Also perhaps the section with `match (range.span, range.min, range.max, range.equal)` could be refactored, what do you think?

### CODE REVIEW

**Comment by Rolv-Apneseth** [comment]
Want me to try fix the `cargo deny` stuff on this branch or would you rather another PR to handle that so it can fix it for both this and #123 after a rebase/merge?

**Comment by jprochazk** [comment]
You can leave that be, I'll fix it later on main

**Comment by Rolv-Apneseth** [comment]
Alright

File: garde_derive/src/syntax.rs [hunk 1]
```diff
@@ -354,27 +365,10 @@ impl Parse for model::RawLength {
...
-                    max = Some(v);
-                }
-                RawLengthArgument::Mode(span, v) => {
-                    if mode.is_some() {
-                        error.maybe_fold(syn::Error::new(span, "duplicate argument"));
-                        continue;
-                    }
-                    mode = Some(v);
-                }
+                RawLengthArgument::Min(span, v) => update_opt!(min, span, v),
```
Comment (jprochazk) [inline garde_derive/src/syntax.rs:368]: Please revert this back to not use a macro.

I personally use macros for implementing a bunch of boilerplate code on some list of things that is subject to change, or to DRY a pattern that is the exact same everywhere and there isn't a neat way to extract it without a macro.
The here code is duplicated 4 times, and when accounting for the added macro, the entire diff is only a few lines of code. It's also unlikely to change much more any time soon.

File: README.md [hunk 2]
```diff
@@ -80,35 +80,38 @@ if let Err(e) = data.validate() {
...
+| credit card  | `#[garde(credit_card)]`                                             | a credit card number                                                                                              | `credit-card`  |
+| phone number | `#[garde(phone_number)]`                                            | a phone number                                                                                                    | `phone-number` |
+| length       | `#[garde(length(<mode>, min=<usize>, max=<usize>, equal=<usize>)]`  | a container with length in `min..=max` or `equal`                                                                 | -              |
+| matches      | `#[garde(matches(<field>))]`                                        | a field matches another field                                                                                     | -              |
+| range        | `#[garde(ra=ge(min=<expr>, max=<expr>, equal=<expr>))]`             | a number in the range `min..=max` or `equal`                                                                      | -              |
```
Comment (jprochazk) [inline README.md:97]:
```suggestion
| range        | `#[garde(range(min=<expr>, max=<expr>, equal=<expr>))]`             | a number in the range `min..=max` or `equal`                                                                      | -              |
```

File: README.md [hunk 3]
```diff
@@ -137,7 +140,7 @@ struct Bar<'a> {
 
 ### Length modes
 
-The `length` rule accepts an optional `mode` argument, which determines what _kind_ of length it will validate.
+The `length` rule accepts an optional `mode` argument, which determines what *kind* of length it will validate.
```
Comment (jprochazk) [inline README.md:143]: Please revert unrelated formatting fixes

File: garde/tests/ui/compile-fail/length_bad_equal_combined.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#![allow(dead_code)]
+
+#[derive(garde::Validate)]
+struct Test<'a> {
+    #[garde(length(min = 1, equal = 10))]
+    field: &'a str,
+}
```
Comment (jprochazk) [inline garde/tests/ui/compile-fail/length_bad_equal_combined.rs:7]: This could use a few additional cases:

```suggestion
struct Test<'a> {
    #[garde(length(min = 1, equal = 10))]
    min_equal: &'a str,
    #[garde(length(max = 1, equal = 10))]
    max_equal: &'a str,
    #[garde(length(min = 1, max = 1, equal = 10))]
    min_max_equal: &'a str,
}
```

File: garde/tests/rules/length.rs [hunk 5]
```diff
@@ -42,9 +42,9 @@ fn length_invalid() {
 
 #[derive(Debug, garde::Validate)]
 struct Exact<'a> {
-    #[garde(length(min = 2, max = 2))]
+    #[garde(length(equal = 2))]
     field: &'a str,
-    #[garde(inner(length(min = 2, max = 2)))]
+    #[garde(inner(length(equal = 2)))]
```
Comment (jprochazk) [inline garde/tests/rules/length.rs:47]: For the sake of backwards compatibility, we should ensure that `length(min=N,max=N)` and `length(equal=N)` have the exact same result. Can you add a test for that?

```rust
#[derive(Debug, garde::Validate)]
struct MinMaxEqual<'a> {
  #[garde(length(min = 2, max = 2))]
  min_max: &'a str,
  #[garde(length(equal = 2))]
  equal: &'a str,
}
```

File: garde/tests/rules/length.rs [hunk 6]
```diff
@@ -42,9 +42,9 @@ fn length_invalid() {
```
Comment (jprochazk) [inline garde/tests/rules/length.rs:42]: `range` now also supports `equal`, so please add the tests you added for `length` to `range` as well.

**Review by jprochazk — CHANGES_REQUESTED** [review]
Thanks for taking a look at this! I have some nitpicks, and this needs a few more tests.

File: README.md (hunk 2 above)
Comment (Rolv-Apneseth) [reply README.md:97]: Good catch sorry about that

File: README.md (hunk 3 above)
Comment (Rolv-Apneseth) [reply README.md:143]: Sure thing, mb

File: garde/tests/ui/compile-fail/length_bad_equal_combined.rs (hunk 4 above)
Comment (Rolv-Apneseth) [reply garde/tests/ui/compile-fail/length_bad_equal_combined.rs:7]: Sure thing

File: garde/tests/rules/length.rs (hunk 5 above)
Comment (Rolv-Apneseth) [reply garde/tests/rules/length.rs:47]: I'll have a look yeah

File: garde/tests/rules/length.rs (hunk 6 above)
Comment (Rolv-Apneseth) [reply garde/tests/rules/length.rs:42]: Ok will do

**Comment by Rolv-Apneseth** [comment]
> Thanks for taking a look at this! I have some nitpicks, and this needs a few more tests.

Thanks for the review - I've made some more changes there, let me know if I've done anything wrong or you need some more tests

**Review by jprochazk — APPROVED** [review]
Thanks!

**Comment by Rolv-Apneseth** [comment]
No problem

**Comment by Rolv-Apneseth** [comment]
As a potential feature, what do you think of a rule for disallowing trailing/leading white space in strings?
